src/mod_ping.erl

```diff
@@ -1,130 +1,101 @@
-%%% General Public License for more details.
-%%%
-%%% You should have received a copy of the GNU General Public License
-%%% along with this program; if not, write to the Free Software
-%%% Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
-%%% 02111-1307 USA
-%%%
+%%% Author  : Piotr Nosek <piotr.nosek@erlang-solutions.com>
+%%% Purpose : XEP-0199 XMPP Ping implementation
+%%% Created : 14 Nov 2019 by Piotr Nosek <piotr.nosek@erlang-solutions.com>
 %%%----------------------------------------------------------------------
 
 -module(mod_ping).
--author('bjc@kublai.com').
+-author('piotr.nosek@erlang-solutions.com').
 
 -behavior(gen_mod).
--behavior(gen_server).
 -xep([{xep, 199}, {version, "2.0"}]).
 -include("mongoose.hrl").
 -include("jlib.hrl").
 
--define(SUPERVISOR, ejabberd_sup).
 -define(DEFAULT_SEND_PINGS, false). % bool()
 -define(DEFAULT_PING_INTERVAL, 60). % seconds
 -define(DEFAULT_PING_REQ_TIMEOUT, 32).
 
-%% API
--export([start_link/2, start_ping/2, stop_ping/2]).
+%% C2S custom info handler
+-export([handle_info/3]).
 
 %% gen_mod callbacks
 -export([start/2, stop/1]).
 
-%% gen_server callbacks
--export([init/1, terminate/2, handle_call/3, handle_cast/2,
-         handle_info/2, code_change/3]).
-
 %% Hook callbacks
 -export([iq_ping/4,
          user_online/4,
          user_offline/5,
          user_send/4,
          user_keep_alive/2]).
 
--record(state, {host = <<"">>,
-                send_pings = ?DEFAULT_SEND_PINGS,
-                ping_interval = ?DEFAULT_PING_INTERVAL,
-                timeout_action = none,
-                ping_req_timeout = ?DEFAULT_PING_REQ_TIMEOUT,
-                timers = dict:new()}).
-
--type state() :: #state{}.
-
 %%====================================================================
-%% API
+%% Info Handler
 %%====================================================================
-start_link(Host, Opts) ->
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:start_link({local, Proc}, ?MODULE, [Host, Opts], []).
 
-start_ping(Host, JID) when JID#jid.lresource =/= <<>> ->
-    %% Guard check above ensures we are not adding a user without a
-    %% resource by accident. If this happens the server ends up in a
-    %% hot loop of iq-error replies.
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:cast(Proc, {start_ping, JID});
-start_ping(_Host, _JID) -> ok.
+handle_info(init, HandlerState, C2SState) ->
+    {add_ping_timer(HandlerState, C2SState), C2SState};
+handle_info(send_ping, HandlerState, C2SState) ->
+    JID = ejabberd_c2s_state:jid(C2SState),
+    Server = ejabberd_c2s_state:server(C2SState),
+    route_ping_iq(JID, Server),
+    {add_ping_timer(HandlerState, C2SState), C2SState};
+handle_info(timeout, HandlerState, C2SState) ->
+    JID = ejabberd_c2s_state:jid(C2SState),
+    Server = ejabberd_c2s_state:server(C2SState),
+    ejabberd_hooks:run(user_ping_timeout, Server, [JID]),
```

Comment:
This needs to be rewritten to use the new `mongoose_hooks` module. Sth like the following:
```erlang
mongoose_hooks:user_ping_timeout(Server, JID)
```